bundle/edu.gemini.itc.shared/src/main/java/edu/gemini/itc/shared/ConfigCreator.java

```diff
@@ -0,0 +1,374 @@
+            this.offsetMessage = offsetMessage;
+        }
+
+        public Config[] getConfig() {
+            return config;
+        }
+    }
+
+    public ConfigCreator(final ItcParameters p) {
+        this.itcParams = p;
+        this.obsDetailParams = p.observation();
+    }
+
+    // create part of config common for all instruments
+    public ConfigCreatorResult createCommonConfig(int numberExposures) {
+        final CalculationMethod calcMethod = obsDetailParams.calculationMethod();
+        ConfigCreatorResult result = new ConfigCreatorResult(new DefaultConfig[numberExposures]);
+        if (numberExposures < 1) {
+            result.addWarning("Warning: Observation overheads cannot be calculated for the number of exposures = 0.");
+        }
+        int numberCoadds = calcMethod.coaddsOrElse(1);
+        String offset = Double.toString(calcMethod.offset());
+        // for spectroscopic observations we consider ABBA offset pattern
+        List<String> spectroscopyOffsets = new ArrayList<>(Arrays.asList("0", offset, offset, "0"));
+        // for imaging observations we consider ABAB offset pattern
+        List<String> imagingOffsets = new ArrayList<>(Arrays.asList("0", offset, "0", offset));
+        List<String> offsetList = new ArrayList<>();
+
+        for (int i = 0; i < (1 + numberExposures / 4); i++) {
+            if (calcMethod instanceof Imaging) {
+                offsetList.addAll(imagingOffsets);
+                result.setOffsetMessage("ABAB dithering pattern");
+            } else if (calcMethod instanceof Spectroscopy) {
+                offsetList.addAll(spectroscopyOffsets);
+                result.setOffsetMessage("ABBA dithering pattern");
+            }
+        }
+
+        for (int i = 0; i < numberExposures; i++) {
+            Config step = new DefaultConfig();
+
+            step.putItem(ExposureTimeKey, obsDetailParams.exposureTime());
+            step.putItem(ObserveTypeKey, InstConstants.SCIENCE_OBSERVE_TYPE);
+            step.putItem(ObserveClassKey, ObsClass.SCIENCE);
+            step.putItem(CoaddsKey, numberCoadds);
+            step.putItem(TelescopePKey, "0");
+            step.putItem(TelescopeQKey, offsetList.get(i));
+
+
+            if (itcParams.telescope().getWFS().equals(GuideProbe.Type.PWFS)) {
+                step.putItem(GuideWithPWFS2Key, StandardGuideOptions.Value.guide);
+            }
+
+            result.getConfig()[i] = step;
+        }
+
+        return result;
+    }
+
+    /**
+     * Instrument-specific configs
+     */
+    public ConfigCreatorResult createGnirsConfig(GnirsParameters gnirsParams, int numExp) {
+        ConfigCreatorResult result = createCommonConfig(numExp);
+
+        for (Config step : result.getConfig()) {
+            step.putItem(ReadModeKey, (gnirsParams.readMode()));
+            step.putItem(InstInstrumentKey, SPComponentType.INSTRUMENT_GNIRS.readableStr);
+            step.putItem(SlitWidthKey, gnirsParams.slitWidth().logValue());
```

Comment:
`Config`, and the sequence model as a whole, was not a great API unfortunately.  There are no guarantees about which keys will be defined. It's impossible to know what the types of the various keys are without checking.  I'd be more comfortable moving all the new API methods that accept a `Config` out of the `spModel` packages and into the ITC itself.  At that point I'd still recommend using the actual typed values from the model (like the `GNIRSParams.SlithWidth` instead of adding `String`s).  That way you can use the APIs they provide instead of doing `String` comparisons.  Nevertheless if it is all internal to the ITC I'm less concerned about it.

Reply:
Thanks for reviewing. I was using Strings originally because I thought I *had* to use Strings for Config (I looked at the Seqexec config indeed). I replaced some of them with typed values after I realized I was wrong, but some were left over. I can try to fix it. I don't think I'll have time for major refactoring like taking all the methods out of spModel. If the whole idea of simulating Config and sequence doesn't seem great anyway, maybe it's now worth the effort too. I have two 1.5 project day left, I'll do what I can.

Reply:
Hi Shane, I have fixed all the Strings in Config and spModel, using typed values instead, in the latest PR. I didn't move anything out of the spModel though. I hope this is ok for testing.